lib/core/networking/websockets/middleware/encryption_middleware.dart

```diff
@@ -0,0 +1,216 @@
+    final List<int> clearText = await CryptoProvider.encryptionAlgo.decrypt(
+      box,
+      secretKey: _sharedKey!,
+    );
+
+    return utf8.decode(clearText);
+  }
+
+  //* Handshaking
+  /// Handle handshaking responses
+  void call(String message) async {
+    // Wait for own key
+    await _ready;
+
+    //? We can not trust the other party
+    // We catch all errors here as we can't trust the incoming data
+    try {
+      // Synchronize access to avoid multiple shared keys
+      await synchronized<void>(() async {
+        // Check if shared key not already established
+        if (_completer.isCompleted) {
+          throw HandshakeException(
+            'Received crypto message after secure connection was established',
+          );
+        }
+
+        // Remove the prefix ($) from the message
+        message = message.substring(1);
+
+        // Determine msg type
+        if (message.startsWith(_CryptoVars.keyResponse) && _session is ClientSession) {
+          await _handleKeyResponse(message, _CryptoVars.keyResponse);
+        } else if (message.startsWith(_CryptoVars.keyRequest) && _session is ServerSession) {
+          // Send own public key
+          _session.send(_CryptoVars.prefix + _CryptoVars.keyResponse + _publicKeyStr);
+
+          // Generate shared key with remote public key
+          await _handleKeyResponse(message, _CryptoVars.keyRequest);
+        } else {
+          throw HandshakeException("Invalid crypto message: '$message'");
+        }
+      });
+    } catch (e) {
+      _session.raise(e);
+    }
+  }
+
+  /// Start to handshakes with the other party
+  Future<void> _initHandshake() async {
+    // Generate our own key pair
+    _ownKeyPair = await CryptoProvider.keyExchangeAlgo.newKeyPair();
+    _publicKeyStr = (await _ownKeyPair.extractPublicKey()).bytes.join('-');
+
+    // Request the other party's public key (only on client side)
+    if (_session is ClientSession) {
+      _session.send(_CryptoVars.prefix + _CryptoVars.keyRequest + _publicKeyStr);
+    }
+  }
+
+  /// Handle the response to our key request
+  Future<void> _handleKeyResponse(String message, String prefix) async {
+    // Extract the public key
+    final String publicKeyStr = message.substring(prefix.length);
+
+    // Check if the public key is valid (format)
+    if (!_CryptoVars.keyRegex.hasMatch(publicKeyStr)) {
+      throw HandshakeException('Received publicKey is invalid');
+    }
+
+    // Construct actual public key
+    final SimplePublicKey publicKey = SimplePublicKey(
+      publicKeyStr.split('-').map((e) => int.parse(e)).toList(),
+      type: KeyPairType.x25519,
+    );
+
+    // Establish shared key
+    _sharedKey = await CryptoProvider.keyExchangeAlgo.sharedSecretKey(
+      keyPair: _ownKeyPair,
+      remotePublicKey: publicKey,
+    );
+
+    // Complete the handshake
+    _completer.complete();
+  }
+}
+
+/// Common shared variables for crypto
+class _CryptoVars {
+  static const String prefix = '\$';
```

Comment:
Please use a raw string instead of escaping the `$` sign
```suggestion
  static const String prefix = r'$';
```